python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -478,50 +495,59 @@ def _get_nodes_allowed_to_terminate(
-        constraint of the worker's node_type.
+        and max_workers constraint of the worker's node_type.
 
-        Returns True exactly when both of the following hold:
+        Returns KeepOrTerminate.keep when both of the following hold:
         (a) The worker's node_type is present among the keys of the current
             config's available_node_types dict.
         (b) Deleting the node would violate the min_workers constraint for that
             worker's node_type.
 
-        Also updates and returns the dictionary of node type counts.
+        Returns KeepOrTerminate.terminate when both the following hold:
+        (a) The worker's node_type is present among the keys of the current
+            config's available_node_types dict.
+        (b) Keeping the node would violate the max_workers constraint for that
+            worker's node_type.
+
+        Return KeepOrTerminate.decide_later otherwise.
+
 
         Args:
             node_type_counts(Dict[NodeType, int]): The non_terminated node
                 types counted so far.
         Returns:
-            bool: True if the node should be kept. False otherwise.
-            Dict[NodeType, int]: Updated node type counts
+            KeepOrTerminate: keep if the node should be kept, terminate if the
+            node should be terminated, decide_later if we could possibly
+            terminate it.
         """
-        new_node_type_counts = copy.deepcopy(node_type_counts)
         tags = self.provider.node_tags(node_id)
         if TAG_RAY_USER_NODE_TYPE in tags:
             node_type = tags[TAG_RAY_USER_NODE_TYPE]
-            if node_type not in self.available_node_types:
-                # The node type has been deleted from the cluster config.
-                # Don't keep the node.
-                return False, new_node_type_counts
-            new_node_type_counts[node_type] += 1
-            min_workers = self.available_node_types[node_type].get(
+
+            min_workers = self.available_node_types.get(node_type, {}).get(
                 "min_workers", 0)
-            max_workers = self.available_node_types[node_type].get(
+            max_workers = self.available_node_types.get(node_type, {}).get(
                 "max_workers", 0)
-            if new_node_type_counts[node_type] <= min(min_workers,
-                                                      max_workers):
-                return True, new_node_type_counts
-
-        return False, new_node_type_counts
+            if node_type not in self.available_node_types:
+                # The node type has been deleted from the cluster config.
+                # Allow terminating it if needed.
+                return KeepOrTerminate.decide_later
+            new_count = node_type_counts[node_type] + 1
+            if new_count <= min(min_workers, max_workers):
+                return KeepOrTerminate.keep
+            if new_count > max_workers and max_workers > 0:
```

Comment:
why do we need `max_workers > 0` ?

Reply:
this is handling max_workers == 0, in which case we assume there is no max.

Reply:
But max_workers==0 does not mean no max, it means just 0 max workers.

Reply:
https://docs.ray.io/en/master/cluster/config.html and in tests, we have 
max_workers = 0 for head node.
I read it as 0 means no max, is it not the case?

Reply:
We often set max_workers of the head node type to 0 if we want to have a head node different than the worker node types. The head node is not counted as a worker and hence max_workers =0 for a head node type means just the head node.

Reply:
> We often set max_workers of the head node type to 0 if we want to have a head node different than the worker node types. The head node is not counted as a worker and hence max_workers =0 for a head node type means just the head node.

fixed.
I think that giving the head node special treatment is not great though.